[ fuse & youngjae ] 우체통 찾기 - 첫번째 PR

silvertae opened the pull request:
# 미리보기
![마을 미리보기2](https://user-images.githubusercontent.com/76121068/229046183-4eae79ae-bf1c-4b8f-a37e-58dd3f6efec0.png)

## 주요작업
- [x] 화면이 열리면, 기획서와 같은 모습의 마을들이 화면에 자동으로 만들어진다.
- [x] 마을의 위치,크기, 갯수 등이 랜덤하게 생성된다. 자유배치가 너무 난해하다면, 위치에 제약조건을 둔다.
- [x] 우체통 갯수가 랜덤하게 생성된다.
- [ ] 빨간색 우체통이 있는 모든 마을을 조사해서 화면에 표시해야 한다.
  - [x] '버튼'UI을 선택하면 빨간 우체통을 가진 마을을 찾는 작업이 시작 된다.
  - [x] 빨간색 우체통을 가진 마을정보를 텍스트로 표현해야 한다
  - [ ] 우체통은 각각 크기가 다르다, 크기 순으로 정렬해서 정보를 표시한다.
  - [ ] 빨간색 우체통을 가진 마을은 2초 뒤에 '빨간색 테두리'로 표시된다.
- [x] 마을안에 마을은 무한대로 존재할 수 있다. 이를 해결할 수 있는 해법을 사용한다
- [x] Node를 탐색하기 위한 API를 직접 개발한다. 

## 기록

1일차: 리드미에 이번 우체통 미션의 학습과 구현 계획을 작성했습니다. 그리고 저희만의 페어 프로그래밍 룰을 정했습니다.

- 이번 미션을 진행하면서 페어 프로그래밍을 어떻게 효율적으로 적용할 수 있을까 고민했습니다.
- 여러 미션 요구사항 중에 우선순위를 어떻게 두고 진행할지에 대해서 고민했습니다.
  (html, css로 미리 마을들 그려놓기 -> querySelector 구현하기 -> 정렬 알고리즘 공부해서 구현해보기 -> 비동기 프로그래밍 학습하기) 

2일차: querySelector와 getElementById의 작동원리를 조사해보고, DFS 알고리즘에 대해 학습하였습니다. 그리고 학습한 내용을 바탕으로 myQuerySelector를 구현했습니다.

- querySelector가 내부적으로 어떻게 동작하는지 찾아보는데 어려움을 겪었습니다.
- DFS 알고리즘을 실제로 어떻게 DOM Tree 탐색에 적용해서 코드를 구현할지를 고민했습니다.

3일차: 랜덤한 여러 마을들을 생성해줄 때 클래스를 사용하면 좋을 것 같다고 생각을 해서, 오전에는 자바스크립트의 class 문법에 대해서 각자 학습하는 시간을 가졌습니다. 그리고 오후에는 학습한 내용을 바탕으로 Village class를 만들고 필요한 메소드들을 추가했습니다.

- class를 사용해서 만들어지는 마을 인스턴스들에 어떤 프로퍼티와 메소드를 넣어줄지 고민했습니다.
- 생성되는 마을의 사이즈와 위치를 어떻게 랜덤하게 정해줄 것인지를 고민했습니다.
- 마을을 겹치지 않게 해주는 조건에 대해서 고민해보았습니다.

4일차: 만들어진 Village class를 이용해서 화면에 겹치지 않는 랜덤한 마을들을 생성해서 그려주는 코드 구현했습니다.

- 마을의 위치를 잡아줄 때 콜스택이 넘치는 문제가 발생해서 이를 어떻게 해결할지 고민 중입니다.
- 마을 안에 마을이 생길 때 점점 작아지는 마을 크기를 어떻게 제한을 둘지 고민했습니다.
- 마을들이 생성될 때 마을들 이름이 알파벳 순서대로 부여되지 않고 중간에 생략되는 문제를 어떻게 해결할지 고민 중입니다.

5일차: 만들어진 마을에 랜덤하게 우체통을 만들고 버튼을 누르면 이를 찾아서 표시해주는 기능을 구현했습니다.

silvertae committed:
Docs: 리드미에 4주차 학습계획서 작성

리드미에 이번 우체통 미션의 학습과 구현 계획을 작성했습니다.
그리고 저희들만의 페어 프로그래밍 룰을 정했습니다.

Youngjaemj committed:
style: eslint 기본 개발환경 설정

eslint를 적용, 설치 및 gitignore 파일 생성 후 node-modules를 추가했습니다.

Youngjaemj committed:
design: html파일 생성 후 마크업

기획서에 나와있는대로 html 구조를 헤더, 메인, 사이드 영역 작성했습니다.

Youngjaemj committed:
design: style.css 파일 생성 후 마을지도를 그림

html로 만들어둔 레이아웃에 기획서에 나와있는
 마을 지도처럼 보이도록 스타일을 입혔습니다.

Youngjaemj committed:
feat: app.js 파일 생성 후 DOM 탐색 함수 추가

app.js 파일을 새로 만들고 DOM Tree에서 원하는 노드 요소를 찾아서 반환해주는
myQuerySelector()를 만들기 위해서

DFS(깊이 우선 탐색) 알고리즘 중 하나인 pre-order traversal 알고리즘을
이용해서 인자로 전달받은 node를 포함해서 모든 자식요소로 존재하는 node들을
찾아서 하나의 배열에 담아서 반환해주는 walkPreOrder() 함수를 찾아서 활용했습니다.

그리고 이 함수를 기반으로 원하는 tag, class, id 이름을 가지는 첫번째
노드를 찾아서 반환해주는 함수 findNodeByTagName, findNodeByClassName,
findNodeById를 추가하였습니다.

silvertae committed:
docs: 지금까지의 학습 및 구현상황 체크리스트 추가

현재까지 진행한 학습 및 구현상황 진행도를 나타내는 체크리스트를 리드미에
추가 하였습니다.

silvertae committed:
feat: myQuerySelector와 myQuerySelectorAll 함수를 구현

기존에 깊이 우선 탐색 알고리즘을 기반으로 만들어둔 함수들을 활용해서
document.querySelector()처럼 css 선택자를 인자로 받아서 일치하는 요소나
요소들이 담긴 배열을 반환하는 myQuerySelector와 myQuerySelectorAll
함수를 구현했습니다.

namse commented on app.js:
```diff
@@ -0,0 +1,88 @@
+}
+
+function findNodeById(nodeId) {
+	const DOMTree = walkPreOrder(document.body);
+	for (const node of DOMTree) {
+		if (node.id === nodeId) {
+			return node;
+		}
+	}
+	return null;
```
퀴즈: 만약 트리에 맨 위에 '우리가 찾으려는 녀석'이 있다면, DomTree 다 도는 것은 시간 낭비가 아닐까!? 어떻게 하면 좋을까?

silvertae committed:
style: html, css 파일들 수정

마을 만들기에 앞서서 html, css 파일들에서 불필요한 중복을 제거하고
몇 가지 수정을 진행했습니다.

silvertae committed:
feat: 마을을 만들때 사용할 Village class 추가

랜덤한 마을들을 생성할 때 사용할 Village class를 추가했습니다. 그리고
class내에 마을의 랜덤한 크기와 위치를 정해줄 때 사용할 여러 속성과
메소드들을 구현해서 추가했습니다.

현재 마을 안에 마을이 생성될 때 자식마을이 부모 마을을 벗어나는 오류가
발생해서 수정 예정입니다.

qkdflrgs commented on class.js:
```diff
@@ -0,0 +1,60 @@
+		const minHeight = this.parentNode.offsetHeight * 0.2 < 50 ? 50 : this.parentNode.offsetHeight * 0.2;
+
+		return Math.floor(Math.random() * (maxHeight - minHeight + 1)) + minHeight;
+	}
```
숫자들의 의미를 이해하는데 어려움이 있었습니다😅 변수명을 만드는 방법도 고민해봤는데 반복되는 숫자들도 아니고 크게 중요하지 않을 수 있겠다는 생각도 드네요...!

qkdflrgs commented on app.js:
DOM API를 어떻게 구현해야될지 고민이 많았는데, 노드 탐색을 통해 각 기능들을 함수단위로 잘 구현하신 것 같습니다👍

qkdflrgs commented on class.js:
```diff
@@ -0,0 +1,60 @@
+	}
+}
+
+const villageY = new Village('Y', 'city');
+villageY.setRandomLocate();
+villageY.draw();
+const villageW = new Village('W', 'Y');
+villageW.setRandomLocate();
+villageW.draw();
+const villageZ = new Village('Z', 'W');
+villageZ.setRandomLocate();
+villageZ.draw();
```
마을을 생성할 때 직접 값을 입력하는 방법보다 데이터를 받아오는 방법을 고민해보면 좋을 것 같습니다!

qkdflrgs commented on readme.md:
```diff
@@ -1 +1,58 @@
+
+\* 진행하면서 수정될 수 있습니다…
+
+## ⏰ 학습 및 구현 계획
+
+좌표값을 계산해서 랜덤한 위치에 마을들을 배치해서 구현하려면 너무 어려울 것 같아서 저희들은 위치들을 미리 지정해 놓고 정렬, 탐색 알고리즘들을 학습하는 것을 목표로 잡았습니다.
+
+- [ ] 마을의 위치를 미리 레이아웃으로 짜 둔다.
+  - [x] 마을의 위치는 position: absolute로 직접 지정해둔다.
+  - [x] 마을의 위치를 id로 구분한다.
+  - [x] 마을안에 다른 마을을 자식 요소로 포함시킨다.
+  - [ ] 마을명 이쁘게 표시해주기
+- [ ] 랜덤한 크기와 갯수의 우체통 렌더링하기
+  - [ ] 우체통을 어떤 요소로 넣을지 고민해보기
+- [ ] DOM APIs 원리를 찾아본다.
+  - [x] querySelector(), querySelectorAll()
+  - [x] getElementById()
+  - [ ] Array.prototype.sort()
+- [x] 탐색 알고리즘 공부하기
+- [ ] myQuerySelector() 구현하기
+  - [x] id로 일치하는 노드 찾는 함수 구현하기
+  - [x] tagName으로 일치하는 노드 찾는 함수 구현하기
+  - [x] class명으로 일치하는 노드 찾는 함수 구현하기
+  - [ ] querySelector()처럼 동작하는 myQuerySelector() 구현하기
+- [ ] myQuerySelectorAll() 구현하기 
+- [ ] 정렬 알고리즘 공부하기
+- [ ] 비동기 프로그래밍 공부하기
```
함께 리뷰를 진행하면서 두 분이 미션해결을 위해 피그마로 그림까지 그려가면서 많은 고민을 하셨다는 것이 느껴졌습니다😀👍

silvertae committed:
feat: app.js에 랜덤한 위치에 마을을 생성해주는 기능 추가

기존에 Village class에 있던 draw() 메소드를 분리해서 새로운 app.js 파일을
만들어서 페이지가 로드되었을 때 화면에 랜덤한 위치에 랜덤한 개수의 마을들을
생성해주는 함수들을 구현하였습니다.

지금은 makeVillage()라는 함수 내부에서 마을 안에 작은 마을들을 랜덤하게
만들어주는 다른 함수를 호출하는 방식으로 일단 구현을 했는데, 중복된
코드들이 많아서 이후에 수정할 예정입니다.

그리고 가끔씩 콜스택이 초과되는 문제랑 마을들 이름이 알파벳 순서대로
부여되지 않고 중간에 생략되는 오류가 발생하고 있습니다.

silvertae committed:
docs: 학습 및 구현계획 체크리스트 추가 및 수정

학습 및 구현계획에 지금까지의 진행상황과 앞으로의 구현 계획을 반영해서
체크리스트를 수정했습니다.

silvertae committed:
design: 메인화면 css 스타일 수정 및 미리보기 추가

기존에 레이아웃만 잡아놨던 메인화면에서 버튼 스타일과 폰트 등의 스타일
수정을 진행했습니다. 그리고 현재까지 구현된 화면을 캡처해서 미리보기로
리드미에 추가했습니다.

silvertae committed:
feat: 만들어진 마을들 중에서 무작위로 골라서 우체통을 만들어주는 기능 구현

처음 페이지가 로드되고 마을이 다 그려지면 makeMailbox() 함수로
그려진 모든 마을들을 불러와서 마을당 25% 정도의 확률로 우체통을
생성해주는 기능을 추가하였습니다.

silvertae committed:
feat: 버튼 클릭시 우체통이 있는 마을들을 찾아서 출력해주는 기능 추가

빨간우체통 찾기 버튼에 클릭 이벤트를 걸어서 존재하는 모든 우체통들을 찾고
그 우체동들이 있는 마을들의 id를 화면에 출력해주는 기능을 구현했습니다.

silvertae committed:
docs: 구현 체크리스트와 미리보기 이미지 수정

구현 체크리스트와 미리보기 이미지를 현재 진행상황에 맞게 수정하였습니다.

namse commented on app.js:
```diff
@@ -0,0 +1,69 @@
+}
+
+function makeVillage() {
+	const maxVillageCount = Math.floor(Math.random() * 3) + 3;
+	let villageCount = 0;
+	while (villageCount < maxVillageCount) {
+		const village = new Village('city');
+		village.setRandomLocate();
+		drawVillage(village);
+		villageCount++;
+		makeVillageInVillage(village.id);
+	}
```
이런 경우에는 for loop 써주시는게 좋을듯? 저만 그런지 모르겠지만, 워낙 for loop에 익숙한 사람들이 많아서, 그들이 더 이해하기 좋을거에요.

silvertae replied:
넵 for문으로 바꿔보겠습니다. 당장 저만 해도 while보다는 for문이 더 익숙하긴 하네요.

namse commented on village.js:
```diff
@@ -0,0 +1,72 @@
+		this.left = 0;
+	}
+
+	getRandomName() {
+		const alphabet = String.fromCharCode(Village.count);
+		Village.count++;
+		return alphabet;
+	}
```
요거 잘 돼요? AZaz 까지 하면 그다음부터는 이상한게 나올거같은 느낌이?

silvertae replied:
마을 개수를 한정해놓고 생각하다보니 그 이후까지는 생각을 못했던 것 같네요. 마을이 늘어나도 이름을 계속해서 지을 수 있는 방법을 생각해보겠습니다.

namse commented on village.js:
```diff
@@ -0,0 +1,72 @@
+
+	getRandomHeight() {
+		const maxHeight = this.parentNode.offsetHeight * 0.5;
+		const minHeight = this.parentNode.offsetWidth * 0.2;
```
요기 오타있네요~ width가 아니라 height여야할듯.

silvertae replied:
헉;; 바로 수정하겠습니다.

namse commented on village.js:
```diff
@@ -0,0 +1,72 @@
+	}
+
+	isOverLap(ref) {
+		if (
+			this.left + this.width < ref.offsetLeft ||
+			this.left > ref.offsetLeft + ref.offsetWidth ||
+			this.top + this.height < ref.offsetTop ||
+			this.top > ref.offsetTop + ref.offsetHeight
+		) {
+			return true;
+		} else {
+			return false;
+		}
```
요런건 if문 없이 return문 하나로 할 수 있어요. 어떻게 할 수 있을까요?

silvertae replied:
`return (this.left + this.width < ref.offsetLeft || this.left > ref.offsetLeft + ref.offsetWidth || this.top + this.height < ref.offsetTop || this.top > ref.offsetTop + ref.offsetHeight)` 로 바꾸면 한 줄로 쓸 수 있을것 같네요. 감사합니다.

namse commented:
퓨즈랑 영재, 고생 많이 했어요.
당장 우리가 짠 코드가 남들보다 적은 것 같고, 많이 구현을 못해본 것 같다고 느껴지겠지만
그러한 차이는 언제든지 존재할 수 밖에 없어요. 또한 그것은 별로 중요하지 않을 수 있답니다.
항상 내가 왜 이 공부를 시작했는지 생각해보고, 내가 그 길로 계속 꾸준히, 재미있게 잘 걸어가고 있는지를 검토하면서 단단한 마음가짐을 가져봅시다.
화이팅! 수고했어요~

silvertae commented:
감사합니다. 공지에 정해진 리뷰어는 다른 분으로 되어있었는데 남세께서 리뷰와 머지 승인을 해주셨네요. 한 주 동안 많은 것을 느낄 수 있었던 페어 프로그래밍이었던 것 같습니다. 계획했던만큼 요구사항을 많이 구현하지는 못했지만, 진행하면서 막힐때마다 영재와 같이 고민하고 구현하면서 더 열심히 하게되었던 것 같아요. 아직 많이 부족하지만 열심히 정진하겠습니다. 영재도 한 주 동안 고생 많으셨어요!

park78951 commented on domAPI.js:
```diff
@@ -0,0 +1,83 @@
+export function myQuerySelector(selector) {
+	const selectorName = selector.substr(1);
+	if (selector.at(0) === '#') {
```
코드를 좀 더 간결하게, 그리고 이해하기 쉽게 짤 수 있는 방법에 대해 고민해보면 좋을 것 같아요. 
at 메서드를 사용하는게 전혀 틀린 것은 아니지만, string의 첫번째 인덱스에 특정 문자열이 포함되어있는지를 확인하는게 주 목적이라면 다른 메서드로 더 가독성 좋고 간결하게 코드를 작성할 수 있지 않을까 생각이듭니다!

silvertae replied:
mdn을 검색해보니 어떤 문자열이 특정 문자로 시작하는지를 확인해서 true 혹은 false를 반환해주는 `String.prototype.startsWith()`라는 메서드가 있네요! `if(startsWith('#'))`라고 작성하면 더 간결하고 가독성이 좋은 코드가 될 것 같습니다. 감사합니다.

park78951 commented on domAPI.js:
```diff
@@ -0,0 +1,83 @@
+}
+
+function walkPreOrder(node, nodes = []) {
+	if (!node) return nodes;
```
사소한 부분이지만, 다른 곳에서 if문을 작성해주신 것을 보면 중괄호를 꼭 넣으셨는데, 실수로 빠뜨린게 아니라면 동료와 컨벤션을 맞추고 시작하면 굉장히 좋아요. 
코드의 양이 많아질 수록 이런 사소한 문법들이 맞춰졌을때 코드를 읽기가 매우 편하더라구요. (Tip: 특히 일할때 이런 컨벤션에 예민하신 분들도 많아요)

silvertae replied:
if 조건을 만족한 후에 실행되는 코드가 짧을 경우에는 종종 중괄호를 생략했었는데, 한 가지 방법으로 통일하는게 더 일관된 방법이었던 것 같습니다. 사소한 문법들도 컨벤션을 정해서 일관된 방식으로 작성할 수 있게 신경을 더 써보겠습니다.

park78951 commented on app.js:
```diff
@@ -0,0 +1,69 @@
+		village.setRandomLocate();
+		drawVillage(village);
+	}
+}
```
`makeVillageInVillageInVillage` 변수명이 재밌어요. 😁 변수명을 보면서 느낀건데 개발 조건에 따라서는 변수명에 `InVillage`가 뒤로 무한대 늘어날 수 있겠다는 생각이 들기도 해요.
makeVillage* 형태의 함수들이 거의 동일한 로직을 수행하고 있는데 중복된 로직을 하나의 함수로 빼서 사용하면 코드의 양도 줄이고 변수명도 이쁘게 바꿀 수 있지 않을까 싶네요!

silvertae replied:
일단 구현하는데 급해서 되는대로 코드를 짜다보니까 마을 안에 마을을 만드는 함수명이 저렇게 되어버린 것 같습니다. 너무 부끄럽네요... 말씀해 주신대로 동일한 로직들이 반복되고 있어 makeVillage 함수를 하나만 두고 그 안에서 조건에 따라 마을 안에 마을을 만드는 함수로 바꿔보겠습니다.

park78951 commented on village.js:
```diff
@@ -0,0 +1,72 @@
+		if (neighborVillages.length === 0) {
+			return;
+		}
+		if (neighborVillages.every(this.isOverLap, this)) {
```
`every` 사용한 건 매우 좋은 것 같아요. 👍 특정 조건에 부합하는지 찾으려고 배열을 무조건 전체 순회할 필요는 없으니까요. 이미 아시겠지만 비슷한 메서드인 `some`도 활용해서 상황에 따라 사용하면 좋을 것 같아요.

그리고 이건 코딩 스타일의 차이일 수도 있지만 `neighborVillages.every(this.isOverLap, this)`의 결과를 변수에 담아서 사용하면 변수명만으로 메서드가 어떤일을 하는지 한눈에 확인 할 수 있어서 가독성이 좋아질 것 같다는 생각도 듭니다~!

silvertae replied:
some, every 메소드를 알게 된지 얼마 안 되어서 이번에 활용해보려고 사용해보았습니다. 😅
조건을 코드로 바로 넣어두기 보다는 변수명으로 역할을 알려주는 것은 매우 좋은 방법 같네요! 
`isVillageLocationNotOverLap` 이라는 이름의 변수면 더 가독성이 좋아질 것 같습니다.

park78951 reviewed (commented):
처음 리뷰하는 건데 함수 분리나 코드가 깔끔해서 놀랐어요. 역시 제가 코드스쿼드 다닐때 보단 다들 실력이 좋으신 것 같습니다.. 👏 날씨 좋은데 너무 공부만 하지 마시고 주말에는 재밌게 노세요! 항상 화이팅입니다.
남세가 동기부여 왕창 해주셔서 저는 조금 쉬엄쉬엄 하시라고..
